### PR TITLE
OSC integration

### PR DESCRIPTION
**Pull request by sinneb**
First version of OSC integration using tinyosc (https://github.com/mhroth/tinyosc). Upon OTTO boot, a threaded OSC server is started on port 9000. The OSC namespace is /OTTO/. For now, the server translates the four rotary encoders (/OTTO/blue /OTTO/yellow etc) and has two pushbuttons for synth and env.  

A simple test-layout for TouchOSC is included.

Fix needed: the OSC thread is killed not nicely, OTTO exits with error.

### CODE REVIEW

File: src/services/osc.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+namespace otto::service::osc {
+  
+  static volatile bool keepRunning = true;
+
+  // handle Ctrl+C
+  static void sigintHandler(int x) {
+    keepRunning = false;
+  }
+
+  void init()
```
Comment (topisani) [inline src/services/osc.cpp]: If it is a non-returning call, call it `run()` instead of `init()`

File: src/services/osc.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "tinyosc.h"
+
+namespace otto::service::osc {
+  
+  static volatile bool keepRunning = true;
```
Comment (topisani) [inline src/services/osc.cpp]: this is what `otto::global::running()` is for

File: src/services/osc.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+  // handle Ctrl+C
+  static void sigintHandler(int x) {
```
Comment (topisani) [inline src/services/osc.cpp]: So this isn't needed either

File: src/services/osc.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+          } else {
+            tosc_message osc;
+            tosc_parseMessage(&osc, buffer, len);
+            //printf("%s\n", tosc_getAddress(&osc));
+            // encoder increase or decrease?
+            int increase_decrease = (int)tosc_getNextFloat(&osc);
+            // which encoder
+            // TODO: Fix ugly prototyping code
+            if(strcmp ("/OTTO/blue",tosc_getAddress(&osc)) == 0) {
+              if(increase_decrease) {
```
Comment (topisani) [inline src/services/osc.cpp]: this should probably  call `service::ui::impl::rotary` directly, do we have access to an increment number, or does it just send one event per "click"

File: src/services/osc.cpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    //// printf("Starting write tests:\n");
+//     int len = 0;
+//     char blob[8] = {0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF};
+//     len = tosc_writeMessage(buffer, sizeof(buffer), "/address", "fsibTFNI",
+//         1.0f, "hello world", -1, sizeof(blob), blob);
+//     tosc_printOscBuffer(buffer, len);
+//     printf("done.\n");
+
+    // register the SIGINT handler (Ctrl+C)
+    signal(SIGINT, &sigintHandler);
```
Comment (topisani) [inline src/services/osc.cpp]: yeah, remove this, and run the loop on `otto::global::running()`, that should do the trick

File: src/services/osc.cpp (hunk 4 above)
Comment (sinneb) [reply src/services/osc.cpp]: It's one event per click

**Comment by sinneb** [comment]
Thanks for your remarks, I made and pushed the changes and fixes. Fixed the crashing thread also, so OTTO exits nicely again now. 

Only issue still open is the integration into boards and the automatic generation of the select statement to handle events. Should we leave that open or resolve it first before integration?

**Comment by topisani** [comment]
Closing this for now, as its not something we want to support right now. 
We have #118 to track this.

**Pull request closed without merging**